packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -315,58 +315,61 @@ fn mem_batch_to_js<'m>(
     batch_id: &str,
     mem: mv8::Object<'m>,
-    metaversion: &Metaversion,
+    persisted: Metaversion,
 ) -> Result<mv8::Value<'m>> {
     let batch = mv8.create_object();
     let batch_id = mv8.create_string(batch_id);
     batch.set("id", mv8::Value::String(batch_id))?;
     batch.set("mem", mem)?;
-    batch.set("mem_version", metaversion.memory())?;
-    batch.set("batch_version", metaversion.batch())?;
+    batch.set("mem_version", persisted.memory())?;
+    batch.set("batch_version", persisted.batch())?;
     Ok(mv8::Value::Object(batch))
 }
 
 fn batch_to_js<'m>(
     mv8: &'m MiniV8,
     mem: &Memory,
-    metaversion: &Metaversion,
+    persisted: Metaversion,
 ) -> Result<mv8::Value<'m>> {
     // TODO: Is `mem.data.len()` different from `mem.size`? (like Vec capacity vs len?)
     let arraybuffer = mv8.create_arraybuffer(mem.data.as_ptr(), mem.size);
-    let batch_id = mem.get_id();
-    mem_batch_to_js(mv8, batch_id, arraybuffer, metaversion)
-}
-
-fn _mut_batch_to_js<'m>(
-    mv8: &'m MiniV8,
-    mem: &mut Memory,
-    metaversion: &Metaversion,
-) -> Result<mv8::Value<'m>> {
-    // TODO: Is `mem.data.len()` different from `mem.size`?
-    let arraybuffer = mv8.create_arraybuffer(mem.as_mut_ptr(), mem.size);
-    let batch_id = mem.get_id();
-    mem_batch_to_js(mv8, batch_id, arraybuffer, metaversion)
+    let batch_id = mem.id();
+    mem_batch_to_js(mv8, batch_id, arraybuffer, persisted)
 }
 
 fn state_to_js<'m, 'a, 'b>(
     mv8: &'m MiniV8,
-    agent_batches: impl Iterator<Item = &'a AgentBatch>,
-    msg_batches: impl Iterator<Item = &'b MessageBatch>,
+    mut agent_batches: impl Iterator<Item = &'a AgentBatch>,
+    mut message_batches: impl Iterator<Item = &'b MessageBatch>,
 ) -> Result<(mv8::Value<'m>, mv8::Value<'m>)> {
     let js_agent_batches = mv8.create_array();
     let js_message_batches = mv8.create_array();
 
     for (i_batch, (agent_batch, message_batch)) in agent_batches
-        .into_iter()
-        .zip(msg_batches.into_iter())
+        .by_ref()
+        .zip(message_batches.by_ref())
         .enumerate()
     {
-        let agent_batch = batch_to_js(mv8, agent_batch.memory(), agent_batch.metaversion())?;
+        let agent_batch = batch_to_js(
+            mv8,
+            agent_batch.batch.segment().memory(),
+            agent_batch.batch.segment().persisted_metaversion(),
+        )?;
         js_agent_batches.set(i_batch as u32, agent_batch)?;
 
-        let message_batch = batch_to_js(mv8, message_batch.memory(), message_batch.metaversion())?;
+        let message_batch = batch_to_js(
+            mv8,
+            message_batch.batch.segment().memory(),
+            message_batch.batch.segment().persisted_metaversion(),
+        )?;
         js_message_batches.set(i_batch as u32, message_batch)?;
     }
+
+    debug_assert!(
+        agent_batches.count() == 0 && message_batches.count() == 0,
+        "Agent batches and message batches needs to have the same size"
+    );
```

Comment:
Is this how the into_iterator works on a zip? Are we sure it consumes only part of message_batches (if agent_batches is smaller) and thus will have a non-zero count? Why don't we just have the debug_assert before the iter and do `agent_batches.len() == message_batches.len()`?

Reply:
`agent_batch` and `message_batch` is an iterator and there is no stable way of ensuring the length of an iterator.
`zip` will stop as soon as one iterator returns `None`, thus if both iterators has no elements left, they had the same length.

Reply:
I'd failed to notice the type, I thought they were slices or something. Cheers for the explanation